controller/backing_image_manager_controller.go

```diff
@@ -0,0 +1,1233 @@
+			c.eventRecorder.Eventf(bim, v1.EventTypeWarning, EventReasonUpdate, "No pod for backing image manager with state %v, will update to state %v", bim.Status.CurrentState, types.BackingImageManagerStateError)
+			bim.Status.CurrentState = types.BackingImageManagerStateError
+		}
+	} else if pod.Spec.NodeName != bim.Spec.NodeID {
+		if bim.Status.CurrentState != types.BackingImageManagerStateError {
+			log.Errorf("Pod node name %v doesn't match backing image manager node ID %v, will update to state %v", pod.Spec.NodeName, bim.Spec.NodeID, types.BackingImageManagerStateError)
+			c.eventRecorder.Eventf(bim, v1.EventTypeWarning, EventReasonUpdate, "Pod node name %v doesn't match backing image manager node ID %v, will update to state %v", pod.Spec.NodeName, bim.Spec.NodeID, types.BackingImageManagerStateError)
+			bim.Status.CurrentState = types.BackingImageManagerStateError
+		}
+	} else if pod.DeletionTimestamp != nil {
+		if bim.Status.CurrentState != types.BackingImageManagerStateError {
+			log.Errorf("Pod deletion timestamp is set for backing image manager with state %v, will update to state %v", bim.Status.CurrentState, types.BackingImageManagerStateError)
+			c.eventRecorder.Eventf(bim, v1.EventTypeWarning, EventReasonUpdate, "Pod deletion timestamp is set for backing image manager with state %v, will update to state %v", bim.Status.CurrentState, types.BackingImageManagerStateError)
+			bim.Status.CurrentState = types.BackingImageManagerStateError
+		}
+	} else {
+		switch pod.Status.Phase {
+		case v1.PodPending:
+			if bim.Status.CurrentState == types.BackingImageManagerStateRunning {
+				log.Errorf("Backing image manager is state %v but the related pod is pending", types.BackingImageManagerStateRunning)
+				c.eventRecorder.Eventf(bim, v1.EventTypeWarning, EventReasonUpdate, "Backing image manager is state %v but the related pod is pending", types.BackingImageManagerStateRunning)
+				bim.Status.CurrentState = types.BackingImageManagerStateError
+			} else {
+				bim.Status.CurrentState = types.BackingImageManagerStateStarting
+			}
+		case v1.PodRunning:
+			// Make sure readiness probe has passed.
+			isReady := true
+			for _, st := range pod.Status.ContainerStatuses {
+				if !st.Ready {
+					isReady = false
+					break
+				}
+			}
+			if !isReady && bim.Status.CurrentState == types.BackingImageManagerStateRunning {
+				log.Errorf("Backing image manager is state %v but the related pod container not ready, will update to state %v", types.BackingImageManagerStateRunning, types.BackingImageManagerStateError)
+				c.eventRecorder.Eventf(bim, v1.EventTypeWarning, EventReasonUpdate, "Backing image manager is state %v but the related pod container not ready, will update to state %v", types.BackingImageManagerStateRunning, types.BackingImageManagerStateError)
+				bim.Status.CurrentState = types.BackingImageManagerStateError
+			} else if bim.Status.CurrentState != types.BackingImageManagerStateRunning {
```

Comment:
If a container of a running bim suddenly becomes not ready, we need to mark the bim as state error first rather than doing nothing, right?

But here we need to modify `} else if bim.Status.CurrentState != types.BackingImageManagerStateRunning {` to `} else if isReady && bim.Status.CurrentState != types.BackingImageManagerStateRunning {` as well.

Reply:
Agree. Thanks